connect/src/receive/replaywithshutdown.rs

```diff
@@ -0,0 +1,113 @@
+use timely::progress::frontier::MutableAntichain;
+
+use timely::dataflow::operators::capture::event::{Event, EventIterator};
+
+/// Replay a capture stream into a scope with the same timestamp.
+pub trait ReplayWithShutdown<T: Timestamp, D: Data> {
+    /// Replays `self` into the provided scope, as a `Stream<S, D>`.
+    fn replay_with_shutdown_into<S: Scope<Timestamp=T>>(self, scope: &mut S, is_running: Arc<AtomicBool>) -> Stream<S, D>;
+}
+
+impl<T: Timestamp, D: Data, I> ReplayWithShutdown<T, D> for I
+where I : IntoIterator,
+      <I as IntoIterator>::Item: EventIterator<T, D>+'static {
+    fn replay_with_shutdown_into<S: Scope<Timestamp=T>>(self, scope: &mut S, is_running: Arc<AtomicBool>) -> Stream<S, D> {
+
+        let mut builder = OperatorBuilder::new("Replay".to_owned(), scope.clone());
+
+        let address = builder.operator_info().address;
+        let activator = scope.activator_for(&address[..]);
+
+        let (targets, stream) = builder.new_output();
+
+        let mut output = PushBuffer::new(PushCounter::new(targets));
+        let mut event_streams = self.into_iter().collect::<Vec<_>>();
+        let mut started = false;
+
+        let mut antichain = MutableAntichain::new();
+
+        builder.build(
+            move |_frontier| { },
+            move |_consumed, internal, produced| {
+
+                if !started {
+                    // The first thing we do is modify our capabilities to match the number of streams we manage.
+                    // This should be a simple change of `self.event_streams.len() - 1`. We only do this once, as
+                    // our very first action.
+                    internal[0].update(Default::default(), (event_streams.len() as i64) - 1);
+                    antichain.update_iter(Some((Default::default(), (event_streams.len() as i64) - 1)).into_iter());
+                    started = true;
+                }
+
+                if is_running.load(Ordering::Acquire) {
+
+                    for event_stream in event_streams.iter_mut() {
+                        while let Some(event) = event_stream.next() {
+                            match *event {
+                                Event::Progress(ref vec) => {
+                                    antichain.update_iter(vec.iter().cloned());
+                                    internal[0].extend(vec.iter().cloned());
+                                },
+                                Event::Messages(ref time, ref data) => {
+                                    output.session(time).give_iterator(data.iter().cloned());
+                                }
+                            }
+                        }
+                    }
+
+                    // Always reschedule `replay`.
+                    activator.activate();
+
+                    output.cease();
+                    output.inner().produced().borrow_mut().drain_into(&mut produced[0]);
+
+                } else {
```

Comment:
Is it possible to also call the frontier cleanup logic once the replayed computation wraps up, e.g. due to end of file or socket connection closing?

Reply:
I would have guessed that we don't need to do this, because when the channel closes all capabilities at the source have already been dropped (i.e. we end up with an empty frontier). Is this not the case? Is the logging logic not cleanly updating progress when it stops?

Reply:
I guess it's a SnailTrail-specific case because I'm manually writing `Event::Progress` events there, so I also have to manually take care of the last `cap` retraction. I'm currently handling this in SnailTrail's `log_pag` with a custom "end of computation" event, but this is of course somewhat brittle; that's why I thought about moving it into the replay logic. I can completely understand if this should remain a special case to be handled by SnailTrail.

Reply:
`I guess it's a SnailTrail-specific case`.

I'm not sure it is: I need to double-check what's the behaviour of `EventWriter` when we use non-Snailtrail logging.

Reply:
We discussed this offline: we believe this should work fine with the default timely logging (via `TIMELY_LOG_ADDR`). @li1 feel free to open an issue if it turns out we missed something.